url7d636e8121_laforge49_JActor2_tgz-pJ8-org_agilewiki_jactor2_core_revisited_EHJ8/src/org/agilewiki/jactor2/core/impl/mtReactors/ReactorMtImpl.java

```diff
@@ -0,0 +1,784 @@
+     * Check if the current message has timed out and poll any child reactors for same.
+     */
+    public void reactorPoll() throws Exception {
+        final double currentTimeMillis = plantScheduler.currentTimeMillis();
+        final double mst = messageStartTimeMillis;
+        if (mst > 0) {
+            int timerMillis = recovery.getMessageTimeoutMillis(this);
+            if (currentRequest != null) {
+                int localTimeout = currentRequest.getMessageTimeoutMillis();
+                if (localTimeout > -1)
+                    timerMillis = localTimeout;
+            }
+            if ((mst + timerMillis) < currentTimeMillis) {
+                recovery.onMessageTimeout(this);
+            }
+        }
+        final Iterator<Closeable> it = getCloseableSet().iterator();
+        while (it.hasNext()) {
+            final Closeable closeable = it.next();
+            if (!(closeable instanceof ReactorMtImpl)) {
+                continue;
+            }
+            final ReactorMtImpl reactor = (ReactorMtImpl) closeable;
+            reactor.reactorPoll();
+        }
+    }
+
+    /**
+     * Returns the CloseableSet. Creates it if needed.
+     *
+     * @return The CloseableSet.
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    protected final Set<Closeable> getCloseableSet() {
+        if (closeables == null) {
+            closeables = Collections.newSetFromMap((Map) new MapMaker()
+                    .concurrencyLevel(1).weakKeys().makeMap());
+        }
+        return closeables;
+    }
+
+    /**
+     * Add a closeable to the list of closeables.
+     *
+     * @param _closeable A closeable to be closed when this ReactorImpl is closed.
+     * @return True when the closeable was added to the list.
+     */
+    @Override
+    public boolean addCloseable(final Closeable _closeable) {
+        if (startedClosing()) {
+            throw new ReactorClosedException(
+                    "call to addCloseable when reactor already started closing: "
+                            + reason);
+        }
+        if (this == _closeable) {
+            return false;
+        }
+        if (!getCloseableSet().add(_closeable)) {
+            return false;
+        }
+        _closeable.asCloseableImpl().addReactor(this);
+        return true;
+    }
+
+    /**
+     * Remove a closeable from the list of closeables.
+     *
+     * @param _closeable The closeable to be removed.
+     * @return True when the closeable was removed.
+     */
+    @Override
+    public boolean removeCloseable(final Closeable _closeable) {
+        if (closeables == null) {
+            return false;
+        }
+        if (!closeables.remove(_closeable)) {
+            return false;
+        }
+        _closeable.asCloseableImpl().removeReactor(this);
+        return true;
+    }
+
+    public boolean isSlow() {
+        return false;
+    }
+
+    public boolean isCommonReactor() {
+        return asReactor() instanceof CommonReactor;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *asReactor* method in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).